app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/hplus/HPlusHandlerThread.java

```diff
@@ -0,0 +1,446 @@
+            HPlusHealthSampleProvider provider = new HPlusHealthSampleProvider(getDevice(), dbHandler.getDaoSession());
+
+            //Insert the Overlays
+            List<HPlusHealthActivityOverlay> overlayList = new ArrayList<>();
+            List<HPlusDataRecord.RecordInterval> intervals = record.getIntervals();
+            for(HPlusDataRecord.RecordInterval interval : intervals){
+                overlayList.add(new HPlusHealthActivityOverlay(interval.timestampFrom, interval.timestampTo, interval.activityKind, deviceId, userId, null));
+            }
+
+            overlayDao.insertOrReplaceInTx(overlayList);
+
+            //Store the data
+            HPlusHealthActivitySample sample = new HPlusHealthActivitySample(
+                    record.timestamp,            // ts
+                    deviceId, userId,            // User id
+                    record.getRawData(),         // Raw Data
+                    record.activityKind,
+                    0,                           // Intensity
+                    ActivitySample.NOT_MEASURED, // Steps
+                    ActivitySample.NOT_MEASURED, // HR
+                    ActivitySample.NOT_MEASURED, // Distance
+                    ActivitySample.NOT_MEASURED  // Calories
+            );
+
+            sample.setProvider(provider);
+
+            provider.addGBActivitySample(sample);
+
+
+        } catch (Exception ex) {
+            LOG.debug(ex.getMessage());
+        }
+
+        mGetSleepTime = Calendar.getInstance();
+        mGetSleepTime.add(Calendar.SECOND, SLEEP_SYNC_PERIOD);
+
+        return true;
+    }
+
+    private void requestNextSleepData() {
+        mGetSleepTime = Calendar.getInstance();
+        mGetSleepTime.add(Calendar.SECOND, SLEEP_RETRY_PERIOD);
+
+        TransactionBuilder builder = new TransactionBuilder("requestSleepStats");
+        builder.write(mHPlusSupport.ctrlCharacteristic, new byte[]{HPlusConstants.CMD_GET_SLEEP});
+        builder.queue(mHPlusSupport.getQueue());
+    }
+
+
+    public boolean processRealtimeStats(byte[] data) {
+        HPlusDataRecordRealtime record;
+
+        try{
+            record = new HPlusDataRecordRealtime(data);
+        } catch(IllegalArgumentException e){
+            LOG.debug((e.getMessage()));
+            return true;
+        }
+
+        if(record.same(prevRealTimeRecord))
+            return true;
+
+        prevRealTimeRecord = record;
+
+        getDevice().setBatteryLevel(record.battery);
+        getDevice().sendDeviceUpdateIntent(getContext());
+
+        //Skip when measuring
+        if(record.heartRate == 255) {
+            getDevice().setFirmwareVersion2("---");
+            getDevice().sendDeviceUpdateIntent(getContext());
+            return true;
+        }
+
+        getDevice().setFirmwareVersion2("" + record.heartRate);
+        getDevice().sendDeviceUpdateIntent(getContext());
+
+        try (DBHandler dbHandler = GBApplication.acquireDB()) {
```

Comment:
OK, you know the details, not me, so maybe my suggestion was rubbish :-)

I thought you would get a bunch a samples in one batch, so you could buffer them during the batch and, once that batch is transferred, commit it to the database.